### PR TITLE
[Android] Adding handlers for auth and certs

### PR DESCRIPTION
**Pull request by marcuspridham**
BUG=XWALK-6147

### CODE REVIEW

**Comment by marcuspridham** [comment]
FYI.  Certificate challenges work fine.  For basic auth, I needed to apply https://github.com/crosswalk-project/crosswalk/pull/3279 for it to work.

**Comment by marcuspridham** [comment]
The current interface exposed by Crosswalk does not expose getKeyTypes or getPrincipals.

https://github.com/crosswalk-project/crosswalk/blob/master/runtime/android/core_internal/src/org/xwalk/core/internal/ClientCertRequestInternal.java

**Comment by marcuspridham** [comment]
Updated commit based on comments.

**Comment by fujunwei** [comment]
ping @axinging

**Comment by fujunwei** [comment]
ping@wuhengzhi

File: platforms/android/src/org/crosswalk/engine/XWalkCordovaResourceClient.java [hunk 1]
```diff
@@ -124,4 +127,33 @@ public void onReceivedSslError(XWalkView view, ValueCallback<Boolean> callback,
...
+    @Override
+    public void onReceivedClientCertRequest(XWalkView view, ClientCertRequest request) {
+        // Check if there is some plugin which can resolve this certificate request
+        PluginManager pluginManager = parentEngine.pluginManager;
+        if (pluginManager != null && pluginManager.onReceivedClientCertRequest(
+                parentEngine.parentWebView, new XWalkCordovaClientCertRequest(request))) {
+            parentEngine.client.clearLoadTimeoutTimer();
+            return;
+        }
+
```
Comment (wuhengzhi) [inline platforms/android/src/org/crosswalk/engine/XWalkCordovaResourceClient.java]: Extra spaces.

File: platforms/android/src/org/crosswalk/engine/XWalkCordovaResourceClient.java [hunk 2]
```diff
@@ -124,4 +127,33 @@ public void onReceivedSslError(XWalkView view, ValueCallback<Boolean> callback,
...
+                parentEngine.parentWebView,
+                new XWalkCordovaHttpAuthHandler(handler), host, realm)) {
+            parentEngine.client.clearLoadTimeoutTimer();
+            return;
+        }
+
+        // By default handle 401 like we'd normally do!
+        super.onReceivedHttpAuthRequest(view, handler, host, realm);
+    }
+
```
Comment (wuhengzhi) [inline platforms/android/src/org/crosswalk/engine/XWalkCordovaResourceClient.java]: Extra spaces.

File: platforms/android/src/org/crosswalk/engine/XWalkCordovaHttpAuthHandler.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+    public XWalkCordovaHttpAuthHandler(XWalkHttpAuthHandler handler) {
+        this.handler = handler;
+    }
+
+    /**
+     * Instructs the XWalkView to cancel the authentication request.
+     */
+    public void cancel () {
+        this.handler.cancel();
```
Comment (wuhengzhi) [inline platforms/android/src/org/crosswalk/engine/XWalkCordovaHttpAuthHandler.java]: ```this.handler``` -> ```handler```, I know it is equivalent, but for the unified code style, suggest to remove ```this.```

File: platforms/android/src/org/crosswalk/engine/XWalkCordovaHttpAuthHandler.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    /**
+     * Instructs the XWalkView to proceed with the authentication with the given credentials.
+     *
+     * @param username
+     * @param password
+     */
+    public void proceed (String username, String password) {
+        this.handler.proceed(username, password);
```
Comment (wuhengzhi) [inline platforms/android/src/org/crosswalk/engine/XWalkCordovaHttpAuthHandler.java]: ditto

**Comment by marcuspridham** [comment]
Thanks.  Updated based on comments and rebased.

File: platforms/android/src/org/crosswalk/engine/XWalkCordovaHttpAuthHandler.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    private final XWalkHttpAuthHandler handler;
+
+    public XWalkCordovaHttpAuthHandler(XWalkHttpAuthHandler handler) {
+        this.handler = handler;
+    }
+
+    /**
+     * Instructs the XWalkView to cancel the authentication request.
+     */
+    public void cancel () {
```
Comment (wuhengzhi) [inline platforms/android/src/org/crosswalk/engine/XWalkCordovaHttpAuthHandler.java]: Extra space before ```(```.

File: platforms/android/src/org/crosswalk/engine/XWalkCordovaHttpAuthHandler.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        handler.cancel();
+    }
+
+    /**
+     * Instructs the XWalkView to proceed with the authentication with the given credentials.
+     *
+     * @param username
+     * @param password
+     */
+    public void proceed (String username, String password) {
```
Comment (wuhengzhi) [inline platforms/android/src/org/crosswalk/engine/XWalkCordovaHttpAuthHandler.java]: ditto.

**Comment by wuhengzhi** [comment]
LGTM for others, thanks.

**Comment by martin-s** [comment]
Is this going to fix issues like?
> [WARNING:http_network_transaction.cc(370)] Blocked proxy response with status 407 to CONNECT request...

**Pull request merged**